docs/src/adr/018adr-inlining.md

```diff
@@ -24,17 +24,41 @@ Since we have recently reworked the inliner in [#1569](https://github.com/inform
 We elect to use the term in a broader sense of "replacing an operator with its definition", in the two ways listed below:
   1. Standard inlining: the instantiation described above
       1. Non-nullary inlining: the instantiation described above, except the inlining skips nullary LET-defined operators
-  2. Pass-by-name inlining: replacing an operator name `A` with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
+  2. Pass-by-name inlining: replacing an operator name `A` used as an argument to a higher-order operator with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
 
-The reason for doing (1) is that, in order to encode `A(x1,...,xn)`, we need to know `e[x1/p1,...,xn/pn]`, not just `e`, since the latter may contain free variables (e.g. `p1` is free in `e`).
+The reason for doing (1) is that, at some point, a rewriting rule would have to generate constraints from `A(e1,...,en)`. 
+To do this, we couldn't just separately encode `body` and `e1,...,en`, because the richness of the data structures allowed in TLA+ makes it difficult to combine independently generated constraints, in cases where the operator parameters are complex expressions (e.g. `e1` is some record with varied nested constructs).
+This is mostly due to the fact that there is no 1-to-1 correspondence between TLA+ objects and SMT datatypes, so encoding object equality is more complicated (which would be needed to express that `ei` instantiates `pi`).
+Therefore we must, no later than at the point of the rewriting rule, know `body[e1/p1,...,en/pn]`. 
 
 The reason for doing (2) is more pragmatic; in order to rewrite expressions which feature any of the higher-order (HO) built-in operators, e.g. `ApaFoldSet(A, v, S)`, we need to know, at the time of rewriting, how to evaluate an application of `A` (e.g. `A(partial, current)` for folding). 
 Performing (2) allows us to make the rewriting rule local, since the definition becomes available where the operator is used, and frees us from having to track scope in the rewriting rules.
 
+### Examples
+Suppose we have an operator 
+```tla
+A(p,q) == p + 2 * q
+```
+
+Then, the result of performing (1) for `A(1, 2)` would be `1 + 2 * 2`. 
+The constant simplification could take the inlined expression and simplify it to `5`, whereas it could not do this across the application boundary of `A(1,2)`.
+
+The result of performing (2) for `ApaFoldSet(A, 0, {1,2,3})` would be
+```tla
+ApaFoldSet(LET A_LOCAL(p,q) == p + 2 * q, 0, {1,2,3})
+```
+
+While this resulting expression isn't subject to any further simplification, notice that it _does_ contain all the required information to fully translate to SMT, unlike `ApaFoldSet(A, 0, {1,2,3})`, which requires external information about `A`.
```

Comment:
I've a general, unrelated question here: why *not* simplify folds like this to a single scalar value?

Reply:
Because you can't in general. E.g.:
```tla
VARIABLE 
  \* @type: Set(Int);
  S

Op == FoldSet(+, 0, S) \* == ??
```
With inlining, the args are always given (even if they're symbolic), but with folds, if the args are symbolic you're SOL, since you need the contents.